GPCS4/Util/UtilSync.h

```diff
@@ -11,38 +11,55 @@
  * protect data structures for a short duration
  * in case the structure is not likely contested.
  */
-class Spinlock 
-{
 
-public:
+#ifdef _M_X64
+#define __UTIL_SYNC_X86__ 1
+#elif _M_IX86
+#define __UTIL_SYNC_X86__ 1
+#elif __x86_64
+#define __UTIL_SYNC_X86__ 1
+#elif __i386
+#define __UTIL_SYNC_X86__ 1
+#endif
 
-	Spinlock() { }
-	~Spinlock() { }
+#ifdef __UTIL_SYNC_X86__
+void pause() {
+    //https://software.intel.com/content/www/us/en/develop/articles/benefitting-power-and-performance-sleep-loops.html
+    //https://www.felixcloutier.com/x86/pause
+#ifdef _MSC_VER
+    __asm pause
+#else
+    asm volatile("pause");
+#endif
+}
+#else
+void pause() {
+}
+#endif
 
-	Spinlock(const Spinlock&) = delete;
-	Spinlock& operator = (const Spinlock&) = delete;
+class Spinlock {
 
-	void lock() 
-	{
-		while (!this->try_lock())
-		{
-			std::this_thread::yield();
-		}
-			
-	}
+public:
+    Spinlock() = default;
+    Spinlock(const Spinlock &) = delete;
+    Spinlock &operator=(const Spinlock &) = delete;
 
-	void unlock() 
-	{
-		m_lock.store(0, std::memory_order_release);
-	}
+    void lock() {
+        uint32_t expected;
+        while (!m_lock.compare_exchange_weak(expected = 0, 1, std::memory_order_release, std::memory_order_consume)) {
```

Comment:
why initialize `expected` when passing it and not on the declaration?

Reply:
> expected: reference to the value expected to be found in the atomic object. Gets stored with the actual value of *this if the comparison fails.

for each iteration, we need to set "expected" to 0

Reply:
ah yes